### PR TITLE
Allow 'non_db' inputs in dynamic namespaces

### PR DESCRIPTION
**Pull request by greschd**
Decides whether a dynamic input should be ``non_db`` by checking if it is a ``Node`` instance.

The potential issue with this could be accidentally creating ``non_db`` inputs. Maybe we should add an `allow_non_db` flag to the `PortNamespace`, to toggle whether ``non_db`` dynamic inputs can be made.

The `PortNamespace` already has a `non_db` attribute, but it's not entirely clear to me what this does. In `_flatten_inputs` at least, the `non_db` attribute isn't checked when expanding `PortNamespace`s.

### CODE REVIEW

**Comment by greschd** [comment]
I will need to test this further -- I think it doesn't work when the workchain is submitted instead of run.

**Comment by greschd** [comment]
Added a test for a dynamic `non_db` input now.

**Comment by sphuber** [comment]
I am not sure if this is the correct solution. When I think about a `PortNamespace`, it being `dynamic` and `non_db` are two separate properties. Dynamic just means that it does not specify up front which inputs it can take, so everything will be accepted (given they match `valid_type` attribute). The `non_db` attribute I think should just mean that none of the inputs within the namespace should be stored. So in `_flatten_inputs` in the branch for the `PortNamespaces` we can check if `non_db` is `True` in which case we simply skip the entire namespace. I also do not think it is a good idea to set the valid types for `Process` in `aiida-core` to be `None`. If a user really wants to be able to pass non database storable inputs in the top level `inputs` namespace, they can manually change the `valid_type` and `dynamic` attributes.

So in conclusion, looking at the title of the PR. I think it would be fine to allow a dynamic `PortNamespace` to take non-db inputs. One can just label the namespace as `non_db=True`. Am I missing a use case here?

**Comment by greschd** [comment]
I totally agree on reverting the ``valid_type = None`` change however.

**Comment by sphuber** [comment]
But if you mark a dynamic `PortNamespace` as `non_db=True` it also does not make sense to pass inputs that should be stored. If that is the use case, one should simply create two namespaces, one that is storable and one that is not. The `non_db` attribute of the `PortNamespace` tells you something about the behavior of the namespace, not its inputs. If it is `non_db=True`, then you are telling that none of the inputs should be stored, so it seems fine that you are not able to pass attributes that should be stored. I mean you can, but I don't understand why one would then be confused if they don't get stored. That's what the `non_db` says exactly

**Comment by greschd** [comment]
The problem I'm having is that I need namespaces which just forward inputs to a sub-workchain. It shouldn't really care if that sub-workchain has `non_db` inputs or not. Otherwise this just becomes a terrible mess when nesting more than one level deep.

I agree that setting `non_db=True` on the namespace should probably mean that all its inputs are `non_db`, but I think the default should be that you can pass both, and it's decided at the individual input level, not at the namespace level. I would also be fine with just removing the `non_db` attribute from the `PortNamespace` altogether.

**Comment by sphuber** [comment]
I think that would make more sense then indeed. To just remove the `non_db` attribute from the `PortNamespace`. I don't think it was added by @muhrin with any particular reason. It was simply done to keep the symmetry with `InputPort`, so both base classes from `plumpy` were updated in `aiida-core` by mixing in `WithNonDb`. Long story short, if we get rid of `WithNonDb` mixin for `PortNamespace` class and we check in the `_flatten_inputs` if the entry is a subclass of `Node` then that would solve your problem correct?

**Comment by greschd** [comment]
I think so, yes. If I'm not mistaken the `non_db` attribute of the `PortNamespace` doesn't actually have any effect right now anyway.

So I just revert the `valid_type` and remove the `WithNonDb` mixin from the `PortNamespace`, correct?

**Comment by sphuber** [comment]
Yes, I guess that would do it

**Comment by greschd** [comment]
As a consequence I think the `sphinxext` should just mark the `non_db` inputs, instead of putting them in a separate paragraph.

**Comment by greschd** [comment]
Alright, I implemented these changes.
